config/xrdsst.yml

```diff
@@ -1,15 +1,17 @@
-api_key:
-- credentials: xrd:secret
-  key: /path/to/ssh_private_key
-  roles:
-  - <SECURITY_SERVER_ROLE_NAME>
-  url: https://localhost:4000/api/v1/api-keys
+ssh_access:
+  admin_credentials: <SECURITTY_SERVER_CREDENTIALS>
+  user: <SSH_USER>
+  private_key: /path/to/ssh_private_key
 security_server:
 - api_key: X-Road-apikey token=<API_KEY>
+  api_key_roles:
```

Comment:
The roles are now listed and placeholders removed, the user does not need to fill the roles any more

Reply:
Great, listing them explicitly seems very reasonable indeed. However, the following consideration was not clarified:

> How about just dropping it completely and filling in at code? Or at the very least, not requiring their definition at server level, but only at root level (which is still kind of annoying, but at least could be feasibly argued to provide forward-compatibility)?

So what were your thoughts about these two points, were there any particular reasons you preferred to keep these at server level instead of root level? Did you see any downsides to removing the element completely (having slept on it, it seems actually better than in the evening)?

Reply:
It would be exceptionally a bad idea to hard-code them in the source code. I agree with the comment that it is unreasonable for the user to type in these roles, but the roles have to remain in the configuration to be changed if required and also to let the user know what kind of roles the api-key has or must have in order to operate the toolkit.

First you commented that the api-key should be per security server, moved it into the security server, now you are recommending to move the roles out to a separate section or lose completely. Yes, I have given it some thought and I disagree with those comments and remain in the opinion that the roles should be kept where the api-key is and list them there.

Reply:
> 
First you commented that the api-key should be per security server, moved it into the security server, now you are recommending to move the roles out to a separate section or lose completely. Yes, I have given it some thought and I disagree with those comments and remain in the opinion that the roles should be kept where the api-key is and list them there.

The ``api_key`` indeed only can be per security server. This is about  ``api_key_roles`` which are for temporary api_keys only and have remained constant. So for persistent API keys used this configuraton does not give any information about API key roles. Both persistent API key and transient API key roles are listed in server status query report.

> It would be exceptionally a bad idea to hard-code them in the source code. I agree with the comment that it is unreasonable for the user to type in these roles, but the roles have to remain in the configuration to be changed if required and also to let the user know what kind of roles the api-key has or must have in order to operate the toolkit.

Could you explain what makes this idea exceptionally bad? I see only forward-compatibility concerns I listed in first comment (which would apply when toolkit becomes unmaintained at one point and X-Road breaks backwards compatibility), but that does not seem exceptionally bad, as the role changes are really slow in X-Road which itself is designed to be backwards-compatible.

Reply:
Hard-coded values in the code is generally a bad idea, there is not much to discuss/explain about it. As the roles are related to api keys and only to that in the sense of configuration, these should be kept close to api key parameters, not in some separate section, which would make it confusing

Reply:
There is much do discuss here, if not blankets, but the case.

So the current standing for specifying transient key roles for 3 security servers, ``as-is-keyroles.yml``

```yaml
# Each temporary SSH key has explicit roles listed at each security server,
# these are available for temporary keys via the administrative credentials.
# If any of these are missing, not all toolkit operations are invokable.
admin_credentials: admuser:admpass
ssh_access:
  user: some_user
  private_key: /path/to/ssh_private_key
security_server:
- name: ss1
  url: https://ss1:4000/api/v1
  api_key: X-Road-apikey token=<API_KEY>
  api_key_roles:
  - XROAD_SYSTEM_ADMINISTRATOR
  - XROAD_SERVICE_ADMINISTRATOR
  - XROAD_SECURITY_OFFICER
  - XROAD_REGISTRATION_OFFICER
  api_key_url: https://localhost:4000/api/v1/api-keys
  # OTHER ss1 configuration
- name: ss2
  url: https://ss2:4000/api/v1
  api_key: X-Road-apikey token=<API_KEY>
  api_key_roles:
  - XROAD_SYSTEM_ADMINISTRATOR
  - XROAD_SERVICE_ADMINISTRATOR
  - XROAD_SECURITY_OFFICER
  - XROAD_REGISTRATION_OFFICER
  api_key_url: https://localhost:4000/api/v1/api-keys
  # OTHER ss2 configuration
- name: ss3
  url: https://ss3:4000/api/v1
  api_key: X-Road-apikey token=<API_KEY>
  api_key_roles:
  - XROAD_SYSTEM_ADMINISTRATOR
  - XROAD_SERVICE_ADMINISTRATOR
  - XROAD_SECURITY_OFFICER
  - XROAD_REGISTRATION_OFFICER
  api_key_url: https://localhost:4000/api/v1/api-keys
  # OTHER ss3 configuration
```

When transient key roles were being defined in application and entirely left out of the configuration, ``code-applied-keyroles.yml``

```yaml
# Roles needed by toolkit are added to temporary key from code,
# these are available for temporary keys via the administrative credentials.
# All toolkit operations are invokable.
admin_credentials: admuser:admpass
ssh_access:
  user: some_user
  private_key: /path/to/ssh_private_key
security_server:
- name: ss1
  url: https://ss1:4000/api/v1
  api_key: X-Road-apikey token=<API_KEY>
  api_key_url: https://localhost:4000/api/v1/api-keys
  # OTHER ss1 configuration
- name: ss2
  url: https://ss2:4000/api/v1
  api_key: X-Road-apikey token=<API_KEY>
  api_key_url: https://localhost:4000/api/v1/api-keys
  # OTHER ss2 configuration
- name: ss3
  url: https://ss3:4000/api/v1
  api_key: X-Road-apikey token=<API_KEY>
  api_key_url: https://localhost:4000/api/v1/api-keys
  # OTHER ss3 configuration
```

And finally, transient key roles defined in the configuration, but only at root level, ``root-keyroles.yml``

```yaml
# Temporary SSH keys have explicit roles listed at root of configuration file,
# again available for temporary keys via the administrative credentials.
# If any of these are missing, not all toolkit operations are invokable.
admin_credentials: admuser:admpass
ssh_access:
  user: some_user
  private_key: /path/to/ssh_private_key
api_key_roles:
  - XROAD_SYSTEM_ADMINISTRATOR
  - XROAD_SERVICE_ADMINISTRATOR
  - XROAD_SECURITY_OFFICER
  - XROAD_REGISTRATION_OFFICER
security_server:
- name: ss1
  url: https://ss1:4000/api/v1
  api_key: X-Road-apikey token=<API_KEY>
  api_key_url: https://localhost:4000/api/v1/api-keys
  # OTHER ss1 configuration
- name: ss2
  url: https://ss2:4000/api/v1
  api_key: X-Road-apikey token=<API_KEY>
  api_key_url: https://localhost:4000/api/v1/api-keys
  # OTHER ss2 configuration
- name: ss3
  url: https://ss3:4000/api/v1
  api_key: X-Road-apikey token=<API_KEY>
  api_key_url: https://localhost:4000/api/v1/api-keys
  # OTHER ss3 configuration
```

So as the key is transient and roles are backwards compatible, after considering this and taking note that ``api_key_roles`` element would most natural next to API keys, I now suggest we go with the code applied keyroles entirely (before I was leaning towards root definition for key roles), as the side effects seem to be strictly about forward-compatibilty, which we can safely ignore, as toolkit code can always take care of them, if these occur in the future.

Reply:
I moved the api_key_roles to the root level